import { Request, Response } from "express";
import { CreateContactsDTO, CustomError } from "../../domain";
import { ContactsService } from "../services/contactsService";

export class ContactsController {

    // El constructor de la clase recibe una instancia del servicio `UserService` que se utilizarÃ¡ para interactuar
    // con la lÃ³gica de negocio para crear usuarios.
    constructor(private readonly contactsService: ContactsService) {}

    // FunciÃ³n privada para manejar los errores. Recibe el error y la respuesta HTTP.
    // Si el error es una instancia de `CustomError`, devuelve un mensaje de error con el cÃ³digo de estado especÃ­fico.
    // Si el error no es una instancia de `CustomError`, se devuelve un error genÃ©rico 500 (Internal Server Error).
    private handleError = (error: unknown, res: Response) => {
        // Verifica si el error es una instancia de CustomError, si es asÃ­, responde con el cÃ³digo de estado y el mensaje adecuado.
        if (error instanceof CustomError) {
            return res.status(error.statusCode).json({ message: error.message });
        };

        // Si el error no es de tipo `CustomError`, muestra el error en la consola y devuelve un error interno genÃ©rico.
        console.log(error);
        return res.status(500).json({ message: "Internal served error ðŸ’©" })
    };

    // MÃ©todo `createUser` para manejar la creaciÃ³n de un nuevo usuario.
    // Este mÃ©todo recibe la solicitud `req` y la respuesta `res` como parÃ¡metros.
    createContact = (req: Request, res: Response) => {
        // Llama a `CreateUsersDTO.create` para validar y crear una instancia de DTO con los datos del cuerpo de la solicitud.
        // Si los datos son incorrectos, se devuelve un error 422 con el mensaje correspondiente.
        const [error, createUsersDto] = CreateContactsDTO.create(req.body);

        if (error) return res.status(422).json({ message: error });

        // Si los datos son vÃ¡lidos, llama al mÃ©todo `createUser` del `UserService` para crear el usuario.
        // En caso de Ã©xito, responde con el cÃ³digo de estado 201 (Creado) y los datos del nuevo usuario.
        this.contactsService.createContact(createUsersDto!)
            .then((data: any) => 
                res.status(201).json(data)
            )
            // Si ocurre un error, se maneja mediante la funciÃ³n `handleError`.
            .catch((error: unknown) => this.handleError(error, res))
    };
}